html/browsers/browsing-the-web/back-forward-cache/resources/helper.sub.js

```diff
@@ -0,0 +1,129 @@
+
+const executorPath =
+  '/html/browsers/browsing-the-web/back-forward-cache/resources/executor.html?uuid=';
+
+// Asserts that the executor `target` is (or isn't, respectively)
+// restored from BFCache. These should be used in the following fashion:
+// 1. Call prepareNavigation() on the executor `target`.
+// 2. Navigate the executor to another page.
+// 3. Navigate back to the executor `target`.
+// 4. Call assert_bfcached() or assert_not_bfcached() on the main test HTML.
+async function assert_bfcached(target) {
+  const status = await getBFCachedStatus(target);
+  assert_implements_optional(status === 'BFCached', 'Should be BFCached');
+}
+
+async function assert_not_bfcached(target) {
+  const status = await getBFCachedStatus(target);
+  assert_implements_optional(status !== 'BFCached', 'Should not be BFCached');
+}
+
+async function getBFCachedStatus(target) {
+  const [loadCount, isPageshowFired] =
+    await target.execute_script(() => [window.loadCount, window.isPageshowFired]);
+  if (loadCount === 1 && isPageshowFired === true) {
+    return 'BFCached';
+  } else if (loadCount === 2 && isPageshowFired === false) {
+    return 'Not BFCached';
+  } else {
+    // This can occur for example when this is called before first navigating
+    // away (loadCount = 1, isPageshowFired = false), e.g. when
+    // 1. sending a script for navigation and then
+    // 2. calling getBFCachedStatus() without waiting for the completion of
+    //    the script on the `target` page.
+    assert_unreached(
+      `Got unexpected BFCache status: loadCount = ${loadCount}, ` +
+      `isPageshowFired = ${isPageshowFired}`);
+  }
+}
+
+// Always call `await remoteContext.execute_script(waitForPageShow);` after
+// triggering to navigation to the page, to wait for pageshow event on the
+// remote context.
+const waitForPageShow = async () => {
+  while (!window.pageShowPromise) {
+    await new Promise(resolve => setTimeout(resolve, 50));
+  }
+  await window.pageShowPromise;
+};
+
+// Run a test that navigates A->B->A:
+// 1. Page A is opened by `params.openFunc(url)`.
+// 2. `params.funcBeforeNavigation` is executed on page A.
+// 3. The window is navigated to page B on `params.targetOrigin`.
+// 4. The window is back navigated to page A (expecting BFCached).
+//
+// Events `params.events` (an array of strings) are observed on page A and
+// `params.expectedEvents` (an array of strings) is expected to be recorded.
+// See `event-recorder.js` for event recording.
+//
+// Parameters can be omitted. See `defaultParams` below for default.
+function runEventTest(params, description) {
+  const defaultParams = {
+    openFunc: url => window.open(url, '_blank', 'noopener'),
+    funcBeforeNavigation: () => {},
+    targetOrigin: originCrossSite,
+    events: ['pagehide', 'pageshow', 'load'],
+    expectedEvents: [
+      'window.load',
+      'window.pageshow',
+      'window.pagehide.persisted',
+      'window.pageshow.persisted'
+    ],
```

Comment:
Should those be mandatory params, especially the expectations? It took me a while to understand what was tested in "SameOrigin", "SameSite" and "CrossSite".

Reply:
I feel it's better to keep these optional. These are the baseline expectations, and if these are mandatory, these params are duplicated in several tests (three times in `events.html`, and more in upcoming tests).

Reply:
ACK.